tests/protocol/test_protocol_util.py

```diff
@@ -138,64 +129,162 @@ def test_detect_wire_protocol_no_dhcp(self, WireProtocol, mock_get_lib_dir, _):
         endpoint_file = protocol_util._get_wireserver_endpoint_file_path()
 
         # Test wire protocol when no endpoint file has been written
-        protocol_util._detect_wire_protocol()
+        protocol_util._detect_protocol()
         self.assertEqual(KNOWN_WIRESERVER_IP, protocol_util.get_wireserver_endpoint())
 
-        # Test wire protocol when endpoint was previously detected
-        protocol_util.clear_protocol()
-        with open(endpoint_file, "w+") as endpoint_fd:
-            endpoint_fd.write("baz.qux")
-
-        protocol_util._detect_wire_protocol()
-        self.assertEqual("baz.qux", protocol_util.get_wireserver_endpoint())
-
         # Test wire protocol on dhcp failure
-        protocol_util.clear_protocol()
         protocol_util.osutil.is_dhcp_available.return_value = True
         protocol_util.dhcp_handler.run.side_effect = DhcpError()
 
-        self.assertRaises(ProtocolError, protocol_util._detect_wire_protocol)
+        self.assertRaises(ProtocolError, protocol_util._detect_protocol)
 
-    @patch("azurelinuxagent.common.protocol.util.MetadataProtocol")
     @patch("azurelinuxagent.common.protocol.util.WireProtocol")
-    def test_get_protocol(self, WireProtocol, MetadataProtocol, _):
+    def test_get_protocol(self, WireProtocol, _):
         WireProtocol.return_value = MagicMock()
-        MetadataProtocol.return_value = MagicMock()
 
         protocol_util = get_protocol_util()
         protocol_util.get_wireserver_endpoint = Mock()
         protocol_util._detect_protocol = MagicMock()
-
-        # Test for wire protocol
         protocol_util._save_protocol("WireProtocol")
 
         protocol = protocol_util.get_protocol()
+
         self.assertEquals(WireProtocol.return_value, protocol)
         protocol_util.get_wireserver_endpoint.assert_any_call()
 
-        # Test to ensure protocol persists
-        protocol_util.get_wireserver_endpoint.reset_mock()
-        protocol_util._save_protocol("MetadataProtocol")
+    @patch('azurelinuxagent.common.conf.get_lib_dir')
+    @patch('azurelinuxagent.common.conf.enable_firewall')
+    def test_get_protocol_wireserver_to_wireserver_update_removes_metadataserver_artifacts(self, mock_enable_firewall, mock_get_lib_dir, _):
+        """
+        This is for testing that agent upgrade from WireServer to WireServer protocol
```

Comment:
I'm sorry I couldn't understand the context of this test, could you please explain briefly this scenario
>agent upgrade from WireServer to WireServer protocol

Thanks!

Reply:
This is testing the normal agent upgrade with both the old agent and new agent using WireServer protocol. However I am testing that MetadataServer certificates are cleaned up during this upgrade. MetadataServer certificates could be leftover because there was a previous upgrade from MetadataServer to WireServer, which does not have the logic for cleaning up MetadataServer certificates.

MDS Agent -> WS Agent (does not clean up MDS certificates, older version) -> WS Agent (version this PR is, has cleanup logic)

Reply:
Ahh now I get it, thanks for adding such a verbose test! :)